src/nunit.analyzers/TestCaseSourceUsage/TestCaseSourceUsesStringAnalyzer.cs

```diff
@@ -53,47 +83,235 @@ private static void AnalyzeAttribute(SyntaxNodeAnalysisContext context)
+            if (positionalArguments.Length < 1)
             {
-                return false;
+                return null;
             }
 
-            foreach (var symbol in context.SemanticModel.LookupSymbols(literal.SpanStart, container: context.ContainingSymbol.ContainingType, name: literal.Token.ValueText))
+            var firstArgumentExpression = positionalArguments[0]?.Expression;
+            if (firstArgumentExpression == null)
+            {
+                return null;
+            }
+
+            // TestCaseSourceAttribute has the following constructors:
+            // * TestCaseSourceAttribute(Type sourceType)
+            // * TestCaseSourceAttribute(Type sourceType, string sourceName)
+            // * TestCaseSourceAttribute(Type sourceType, string sourceName, object?[]? methodParams)
+            // * TestCaseSourceAttribute(string sourceName)
+            // * TestCaseSourceAttribute(string sourceName, object?[]? methodParams)
+            if (firstArgumentExpression is TypeOfExpressionSyntax typeofSyntax)
+            {
+                var sourceType = context.SemanticModel.GetSymbolInfo(typeofSyntax.Type).Symbol as INamedTypeSymbol;
+                return ExtractElementsInAttribute(context, sourceType, positionalArguments, 1);
+            }
+            else
+            {
+                var sourceType = context.ContainingSymbol.ContainingType;
+                return ExtractElementsInAttribute(context, sourceType, positionalArguments, 0);
+            }
+        }
+
+        private static SourceAttributeInformation ExtractElementsInAttribute(
+            SyntaxNodeAnalysisContext context,
+            INamedTypeSymbol sourceType,
+            ImmutableArray<AttributeArgumentSyntax> positionalArguments,
+            int sourceNameIndex)
+        {
+            if (sourceType == null)
+            {
+                return null;
+            }
+
+            SyntaxNode syntaxNode = null;
+            string sourceName = null;
+            bool isStringLiteral = false;
+            if (positionalArguments.Length > sourceNameIndex)
+            {
+                var syntaxNameAndType = GetSyntaxStringConstantAndType(context, positionalArguments, sourceNameIndex);
+
+                if (syntaxNameAndType == null)
+                {
+                    return null;
+                }
+
+                syntaxNode = syntaxNameAndType.Item1;
+                sourceName = syntaxNameAndType.Item2;
+                isStringLiteral = syntaxNameAndType.Item3;
+            }
+
+            int? numMethodParams = null;
+            if (positionalArguments.Length > sourceNameIndex + 1)
+            {
+                numMethodParams = GetNumberOfParametersToMethod(positionalArguments[1]);
+            }
+
+            return new SourceAttributeInformation(sourceType, sourceName, syntaxNode, isStringLiteral, numMethodParams);
+        }
+
+        private static Tuple<SyntaxNode, string, bool> GetSyntaxStringConstantAndType(

```

Comment:
Can we use ValueTuple here?
`(SyntaxNode syntaxNode, string sourceName, bool isLiteral)`

Reply:
Yes we can